docs: document overriding default middleware by name (#4701)
langchain-ai/docs: src/oss/deepagents/customization.mdx

| [`backend=`](#backends) | Filesystem backend (StateBackend by default) |
| [`permissions=`](/oss/deepagents/permissions) | Path-level access control for the filesystem |
| [`subagents=`](#subagents) | Custom subagents for delegated tasks |

> | [`middleware=`](#middleware) | Extra middleware merged into the [default stack](#default-stack-main-agent); an instance whose `.name` matches a default replaces it in place, anything else lands after the last core middleware entry and before the profile/prompt-caching/memory tail |

Reviewer: ```suggestion
| [`middleware=`](#middleware) | Extra middleware merged into the [default stack](#default-stack-main-agent); an instance whose `.name` matches a default replaces it in place, anything else lands after the last core middleware entry and before the profile, prompt-caching, and memory |
```

5. @[`SummarizationMiddleware`]: Condenses message history to stay within context limits when conversations grow long (via @[create_summarization_middleware]).
6. @[`PatchToolCallsMiddleware`]: Repairs dangling tool calls in message history when a run resumes after an interruption or receives malformed tool-call arguments. Runs **before** Anthropic prompt caching and the tail stack below.
7. @[`AsyncSubAgentMiddleware`]: Only when you configure async subagents.

> 8. **Your middleware argument**: Optional middleware you pass as the `middleware` argument is merged here (after Patch, before the tail stack)—an instance whose `.name` matches one of the defaults above replaces that default in place instead of duplicating it; anything else lands here. See [Override a default middleware instance](#override-a-default-middleware-instance).

Reviewer: ```suggestion
8. **Your middleware argument**: Optional middleware you pass as the `middleware` argument is merged after Patch but before the rest of the stack. An instance whose `.name` matches one of the defaults above replaces that default in place instead of duplicating it; anything else lands here. See [Override a default middleware instance](#override-a-default-middleware-instance).
```

If you must use mutation in custom middleware, consider what happens when subagents, parallel tools, or concurrent agent invocations run at the same time.
</Warning>

### Override a default middleware instance

:::python
<Note>

> Overriding a default middleware by matching `.name` requires `deepagents` 0.7a3 versions or newer.

Reviewer 2: ```suggestion
Overriding a default middleware by matching `.name` requires `deepagents>=0.7.0a3`.
```

</Note>

Pass a middleware instance whose `.name` matches an entry in the [default stack](#default-stack-main-agent), such as @[`SummarizationMiddleware`], to replace that default in place instead of appending a duplicate. Any middleware you pass whose `.name` does **not** match a default is not replaced, it lands after the last core middleware entry and before the profile, prompt-caching, and memory. See [Default stack (main agent)](#default-stack-main-agent) for the full ordering.

> ```python

Reviewer 2: As we accumulate in-demand use cases (e.g., updating prompt cache TTL), adding them as accordion sections here as examples could be nice.

Author: Added

If you must use mutation in custom middleware, consider what happens when subagents, parallel tools, or concurrent agent invocations run at the same time.
</Warning>

### Override a default middleware instance

:::python
<Note>

> Overriding a default middleware by matching `.name` requires `deepagents` 0.7 alpha versions.

Reviewer: Presumably it'll work in later versions too
```suggestion
Overriding a default middleware by matching `.name` requires `deepagents` 0.7 alpha versions or newer.
```

</Note>

> Pass a middleware instance whose `.name` matches an entry in the [default stack](#default-stack-main-agent), such as @[`SummarizationMiddleware`], to replace that default in place instead of appending a duplicate. Any middleware you pass whose `.name` does **not** match a default is not replaced, it lands after the last core middleware entry and before the profile/prompt-caching/memory tail. See [Default stack (main agent)](#default-stack-main-agent) for the full ordering.

Reviewer: ```suggestion
Pass a middleware instance whose `.name` matches an entry in the [default stack](#default-stack-main-agent), such as @[`SummarizationMiddleware`], to replace that default in place instead of appending a duplicate. Any middleware you pass whose `.name` does **not** match a default is not replaced, it lands after the last core middleware entry and before the profile, prompt-caching, and memory. See [Default stack (main agent)](#default-stack-main-agent) for the full ordering.
```
